File: app/src/main/java/com/pr0gramm/app/ui/upload/SimilarImageView.java
package com.pr0gramm.app.ui.upload;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.ColorDrawable;
import android.net.Uri;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.google.common.collect.ImmutableList;
import com.pr0gramm.app.Dagger;
import com.pr0gramm.app.HasThumbnail;
import com.pr0gramm.app.R;
import com.pr0gramm.app.services.UriHelper;
import com.squareup.picasso.Picasso;

import java.util.List;

/**
 */
public class SimilarImageView extends RecyclerView {
    public SimilarImageView(Context context) {
        super(context);
        init();
    }

    public SimilarImageView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public SimilarImageView(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
        setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
    }

    public void setThumbnails(Iterable<HasThumbnail> thumbnails) {
        setAdapter(new ThumbnailAdapter(ImmutableList.copyOf(thumbnails)));
    }

    private class ThumbnailAdapter extends Adapter<ThumbnailViewHolder> {
        private final List<HasThumbnail> thumbnails;
        private final Picasso picasso = Dagger.appComponent(getContext()).picasso();

        ThumbnailAdapter(List<HasThumbnail> thumbnails) {
            this.thumbnails = thumbnails;
        }

        @Override
        public void onBindViewHolder(ThumbnailViewHolder holder, int position) {
            HasThumbnail thumb = thumbnails.get(position);

            Uri imageUri = UriHelper.Companion.of(getContext()).thumbnail(thumb);
            picasso.load(imageUri)
                    .config(Bitmap.Config.RGB_565)
                    .placeholder(new ColorDrawable(0xff333333))
                    .into(holder.image);
        }

        @Override
        public ThumbnailViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.thumbnail, parent, false);

            return new ThumbnailViewHolder(view);
        }

        @Override
        public int getItemCount() {
            return thumbnails.size();
        }
    }

    private static class ThumbnailViewHolder extends ViewHolder {
        final ImageView image;

        public ThumbnailViewHolder(View itemView) {
            super(itemView);
            image = (ImageView) itemView;
        }
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/upload/SimilarImageView.kt
package com.pr0gramm.app.ui.upload

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.drawable.ColorDrawable
import android.util.AttributeSet
import androidx.recyclerview.widget.RecyclerView
import com.pr0gramm.app.api.pr0gramm.Api
import com.pr0gramm.app.api.pr0gramm.asThumbnail
import com.pr0gramm.app.databinding.ThumbnailBinding
import com.pr0gramm.app.services.UriHelper
import com.pr0gramm.app.ui.MainActivity
import com.pr0gramm.app.ui.views.InjectorViewMixin
import com.pr0gramm.app.ui.views.SingleTypeViewBindingAdapter
import com.pr0gramm.app.ui.views.instance
import com.pr0gramm.app.util.observeChange
import com.squareup.picasso.Picasso

/**
 */
class SimilarImageView @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : RecyclerView(context, attrs, defStyleAttr), InjectorViewMixin {

    private val picasso: Picasso by instance()

    init {
        layoutManager = androidx.recyclerview.widget.LinearLayoutManager(context, androidx.recyclerview.widget.LinearLayoutManager.HORIZONTAL, false)
    }

    var items: List<Api.Posted.SimilarItem> by observeChange(listOf()) {
        adapter = itemAdapter(items)
    }

    private fun itemAdapter(items: List<Api.Posted.SimilarItem>): SingleTypeViewBindingAdapter<Api.Posted.SimilarItem, ThumbnailBinding> {
        return SingleTypeViewBindingAdapter(ThumbnailBinding::inflate, items) { views, item ->
            val imageUri = UriHelper.of(context).thumbnail(item.asThumbnail())
            picasso.load(imageUri)
                    .config(Bitmap.Config.RGB_565)
                    .placeholder(ColorDrawable(0xff333333.toInt()))
                    .into(views.bindings.root)

            views.bindings.root.setOnClickListener {
                handleItemClicked(item)
            }
        }
    }

    private fun handleItemClicked(item: Api.Posted.SimilarItem) {
        context.startActivity(MainActivity.openItemIntent(context, item.id).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK
        })
    }
}
